plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/LibraryClasspathContainerInitializer.java

```diff
@@ -0,0 +1,100 @@
+import org.eclipse.jdt.core.JavaCore;
+
+/**
+ * {@link ClasspathContainerInitializer} implementation that resolves containers for App Engine
+ * libraries.
+ * <p>
+ * The container path is expected to be in the form of
+ * &lt;value of {@link Library#CONTAINER_PATH_PREFIX}&gt;/&lt;library ID&gt;
+ */
+public class LibraryClasspathContainerInitializer extends ClasspathContainerInitializer {
+
+  @Inject
+  private LibraryClasspathContainerSerializer serializer;
+  @Inject
+  private ILibraryClasspathContainerResolverService resolverService;
+  private String containerPathPrefix = Library.CONTAINER_PATH_PREFIX;
+
+  public LibraryClasspathContainerInitializer() {
+  }
+
+  @VisibleForTesting
+  LibraryClasspathContainerInitializer(String containerPathPrefix,
+                                       LibraryClasspathContainerSerializer serializer,
+                                       ILibraryClasspathContainerResolverService resolverService) {
+    this.containerPathPrefix = containerPathPrefix;
+    this.serializer = serializer;
+    this.resolverService = resolverService;
+  }
+
+  @Override
+  public void initialize(IPath containerPath, IJavaProject project) throws CoreException {
+    if (containerPath.segmentCount() != 2) {
+      throw new CoreException(StatusUtil.error(this, Messages.getString("ContainerPathNotTwoSegments",
+                                                                        containerPath.toString())));
+    }
+    if (!containerPath.segment(0).equals(containerPathPrefix)) {
+      throw new CoreException(StatusUtil.error(this,
+                                               Messages.getString("ContainerPathInvalidFirstSegment",
+                                                                  containerPathPrefix,
+                                                                  containerPath.segment(0))));
+    }
+    try {
+      LibraryClasspathContainer container = serializer.loadContainer(project, containerPath);
+      if (container != null && jarPathsAreValid(container)) {
+        JavaCore.setClasspathContainer(containerPath,
+                                       new IJavaProject[] {project},
+                                       new IClasspathContainer[] {container},
+                                       new NullProgressMonitor());
+      } else {
+        resolverService.resolveContainer(project, containerPath, new NullProgressMonitor());
+      }
+    } catch (IOException ex) {
+      throw new CoreException(StatusUtil.error(this, Messages.getString("LoadContainerFailed"), ex));
+    }
+  }
+
+  private static boolean jarPathsAreValid(LibraryClasspathContainer container) {
+    IClasspathEntry[] classpathEntries = container.getClasspathEntries();
+    for (int i = 0; i < classpathEntries.length; i++) {
+      IClasspathEntry classpathEntry = classpathEntries[i];
+      if (!classpathEntry.getPath().toFile().exists()
+          || (classpathEntry.getSourceAttachmentPath() != null
```

Comment:
If the entry does not have a source attachment `classpathEntry.getSourceAttachmentPath()` will be null, therefore the second part of the expression will be false, which means the whole expression evaluates to whatever the first part (`classpathEntry.getPath().toFile().exists()`) evaluates to. If the binary jar file exists, the conditional expression will evaluate to false, leading the the `else` branch to be executed.
I've added verification to the corresponding test method `com.google.cloud.tools.eclipse.appengine.libraries.LibraryClasspathContainerInitializerTest.testInitialize_ifSourcePathIsNullContainerIsNotResolvedAgain()`